### DISCUSSION TITLE
Weird Discord bug after mass deleting messages that Undiscord isn't taking into consideration

### DISCUSSION THREAD

**Discussion by Liggliluff** [victornpb/undiscord#450] [3. Ideas]
In short. After mass-deleting messages, Discord search will act weirdly. If you search for that user, you will get some blank pages, but if you continue to scroll, you will find actual messages. Undiscord seems to still be able to estimate the total number of messages, but if the first page or more are blank, then it will fail to delete any messages and just give up.

But if you set the "before" date to a date that makes the first page have a valid message show up in Discord search, then Undiscord will find these messages and delete them. But after deleting the messages of the first page, and make a new search, Discord must at this poit return a blank first page again.

```
Started at 2023-03-02 02:44:09
authorId="REDACTED" guildId="REDACTED" channelId="REDACTED" minId="REDACTED" maxId="REDACTED" hasLink=false hasFile=false
Total messages found: 983	(Messages in current page: 25, To be deleted: 25, System: 0)	offset: 0
Delete delay: 1000ms, Search delay: 100ms	Last Ping: 1989ms, Average Ping: 1989ms
Estimated time remaining: 0h 49m 2s
Waiting for your confirmation...
OK
0.10% (1/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.20% (2/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.31% (3/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.41% (4/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.51% (5/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.61% (6/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.71% (7/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.81% (8/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.92% (9/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.02% (10/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.12% (11/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.22% (12/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.32% (13/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.42% (14/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.53% (15/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.63% (16/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.73% (17/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.83% (18/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.93% (19/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.03% (20/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.14% (21/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.24% (22/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.34% (23/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.44% (24/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.54% (25/983)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 983	(Messages in current page: 0, To be deleted: 0, System: 0)	offset: 0
Delete delay: 1000ms, Search delay: 100ms	Last Ping: 540ms, Average Ping: 439ms
Estimated time remaining: 0h 23m 38s
Ended because API returned an empty page.
Ended at 2023-03-02 02:44:46! Total time: 0h 0m 37s
Delete delay: 1000ms, Search delay: 100ms	Last Ping: 540ms, Average Ping: 439ms
Rate Limited: 0 times. Total time throttled: 0h 0m 0s.
Deleted 25 messages, 0 failed.
```

This search was done with the before date set to 2017-11-04 21:00, and after deleting 25/983 messages, it fails. The last message is dated 2017-11-03 14:04:44. If I retry without changing the date, it will find 983 messages, but 0 to delete, and will stop. If I update the before date to 2017-11-03 14:05, it works again.

```
Started at 2023-03-02 02:51:22
authorId="REDACTED" guildId="REDACTED" channelId="REDACTED" minId="REDACTED" maxId="REDACTED" hasLink=false hasFile=false
Total messages found: 959	(Messages in current page: 24, To be deleted: 24, System: 0)	offset: 0
Delete delay: 1000ms, Search delay: 100ms	Last Ping: 1188ms, Average Ping: 1188ms
Estimated time remaining: 0h 35m 2s
Waiting for your confirmation...
OK
0.10% (1/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.21% (2/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.31% (3/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.42% (4/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.52% (5/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.63% (6/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.73% (7/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.83% (8/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
0.94% (9/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.04% (10/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.15% (11/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.25% (12/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.36% (13/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.46% (14/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.56% (15/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.67% (16/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.77% (17/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.88% (18/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
1.98% (19/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.09% (20/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.19% (21/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.29% (22/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.40% (23/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
2.50% (24/959)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	REDACTED
Searching next messages in 100ms...	
Total messages found: 959	(Messages in current page: 0, To be deleted: 0, System: 0)	offset: 0
Delete delay: 1000ms, Search delay: 100ms	Last Ping: 791ms, Average Ping: 424ms
Estimated time remaining: 0h 22m 50s
Ended because API returned an empty page.
Ended at 2023-03-02 02:52:04! Total time: 0h 0m 42s
Delete delay: 1000ms, Search delay: 100ms	Last Ping: 791ms, Average Ping: 424ms
Rate Limited: 0 times. Total time throttled: 0h 0m 0s.
Deleted 24 messages, 0 failed.
```

The issue is clearly that Discord has for some reason stopped considering deleted messages to be deleted. Undiscord need a way to work around this and still delete messages. It only tries to delete page 1 and assumes Discord has updated the search, which isn't always the caase.

So there are two "solutions", and both wouldn't be bad to implement:

1. Cache multiple pages and not just page 1 before starting to delete. Certainly caching up to 10 pages would work fine? Perhaps Discord has a rate limit here too.
2. After deleting all cached pages (1 or however more there are), take the date-time of the last message deleted and set that as the temporary before date. Next search will be done from this date-time instead.

**Comment by Liggliluff** [comment]
Since Undiscord assumes there are no messages if page 1 is empty, which is false, and there's no interest in caching more than one page for deletion. Then the best solution I have right now is:

1. If there are more than 1 page, and page 1 is empty, continue until the first message is found. If no message is found after reaching the last page, no messages can be deleted.
2. Delete all messages found.
3. Before searching again, set the before date to the date of the last message deleted (but round the value up, so 15:23:05 becomes 15:24, so a message at 15:23:02 isn't missed). This will skip all the pages that are already confirmed to be empty.
4. Repeat from step 1, even after updating the before date, there could still be an empty first page.

**Reply by ahsanm006** [reply]
good workaround 
any updates on when this problem might get fixed?
still got sht ton of messages to get rid of :(

**Reply by Liggliluff** [reply]
@abbydiode that's not a viable solution since I have waited days for the first page to be re-populated. It did seem to fix itself after I posted a message and waited several hours. So Discord won't update the search if the server is inactive. Not always viable.

Still, I do think Undiscord should take this into consideration. An empty first page, with additional pages, does not mean there are no messages. It would give a more reliable tool that goes around this Discord bug.

**Comment by victornpb** [comment]
thank you for the detailed issue, unfortunately these edge cases are hard to test for since I no longer have anything that old to delete, and I can't send messages in the past.

 I have test accounts but I can only send fresh messages, in order to delete, if you still old have messages and you volunteer to be a test subject, we can try something.

I have a new update in the works right now, the code has been tidy up significantly, should be easier to test for a new end condition.

That said I haven't thought too much about your suggestion but I will try to take a careful look at it later today if I can.

**Reply by Liggliluff** [reply]
While I do understand it's hard to test and it's a niche situation. Personally this is what I do use the script for, to delete old archived messages. It's a mostly inactive server, and this issue started to occur after deleting thousands of messages while the server was inactive for I think at least 2 days, which isn't much for a slow or even abandoned server, of which you want to remove your messages for private reasons.

I don't think it has to do with the age of the messages, and I suspect it's just that if a server is inactive for 2 days or so, the search isn't being re-cached, and this issue isn't really noticeable until you delete thousands of messages (with is what this script is for, right?).

Sometimes you don't even have the option to update the server, for example if it's an old archived server where writing is disabled, or it's a DM with someone I no longer want to be in contact with and don't want to randomly alert them, and if someone were to block me then I can also not write in the DMs.

The last (my message) was sent at 2023-03-02 03:20 UTC+1, and I still got several thousands of messages to delete. So we'll see if the issue appears again. I do also wonder if it would speed up the searching after completing a page or in some other way improve it, if the "before date" was updated to be the time of the last message, since all prior messages are confirmed deleted anyway. It doesn't solve the issue where it stops when it finds an empty page, but it will reduce the likelihood of finding an empty page.

----

5 minutes after writing this reply and:
```
Ended because API returned an empty page.
Ended at 2023-03-02 16:54:24 [UTC+1]! Total time: 2h 42m 5s
Deleted 3147 messages, 1 failed. [out of 7000+]
```
So it took 13 h 34 min before it's acting up again. Although rerunning the script does make it work again. So even though the search works, having it not just stop after finding one empty page still helps with random bugs too.

For example, imagine letting it run to delete messages for hours while you're away. You're back after 24 hours, and it has failed 3 hours in because Discord randomly returned an empty page once. 21 hours wasted. It really shouldn't give up this easily. It really should check more than one page, if possible.

**Reply by victornpb** [reply]
> this script is for, right?

i agree. I will try my best to make this work.


currently there's already a workaround when a person have dozens of system messages "person have started a call..." and they pile up at the bottom filling an entire page of messages that can't be deleted. so in that case we increase the offset and keep looking for messages pages forward.

on that stuck channel, if you use the discord app search what the results look like? it shows how many pages at the bottom?

**Reply by Liggliluff** [reply]
When I searched for my username with no other filters inside Discord itself, when it "found" the deleted messages as real messages. The whole first page was empty in Discord, but the page navigation was still available. I could navigate to page 2 and 3 as usual, which were also empty. Page 4 had 2 messages (and presumably 23 deleted messages) and page 5 and onwards appears as normal.

So these phantom deleted messages won't show up, but they still take up the place of a real message.

**Reply by ahsanm006** [reply]
yes , I have seen that before too.
and btw, any update to delete messages in threads/forums too?
when I try to, it says -" Error searching messages, API responded with status 403!
	{"message":"Missing Access","code":50001} "
